### PR TITLE
Approval process

### PR DESCRIPTION
**Pull request by apelisse**
Implements https://github.com/kubernetes/contrib/issues/1389

Please do not remove the `do-not-merge` label, unless you know what you're doing.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/2076)

### CODE REVIEW

**Comment by deads2k** [comment]
@apelisse unassigned myself to let the bot try to assign again.  I haven't written or reviewed in this area before, is there someone more familiar to assign it to?

**Comment by eparis** [comment]
We shouldn't be touching `vendor`, should we?

**Comment by apelisse** [comment]
@eparis I really need your script to check that, I keep forgetting ;-)

**Comment by luxas** [comment]
I guess @eparis is a good reviewer for this PR :)

**Comment by grodrigues3** [comment]
cc @bgrant0607

**Comment by apelisse** [comment]
This pull-request is a merge of multiple pull-requests that have been reviewed extensively already:

- https://github.com/kubernetes/contrib/pull/1923
- https://github.com/kubernetes/contrib/pull/2002
- https://github.com/kubernetes/contrib/pull/2018

**Comment by grodrigues3** [comment]
@thockin @philips

File: mungegithub/features/repo-updates.go
Comment (philips) [inline mungegithub/features/repo-updates.go]: nit: double space

File: mungegithub/features/repo-updates.go
Comment (philips) [inline mungegithub/features/repo-updates.go]: nit: space between // and if

**Comment by thockin** [comment]
I have zero context - what do I need to do/know?

File: mungegithub/mungers/approval-handler.go
Comment (philips) [inline mungegithub/mungers/approval-handler.go]: latestNotification will initialize to nil. This case can be deleted and only the inverse needs to be handled.

File: mungegithub/mungers/approval-handler.go
Comment (philips) [inline mungegithub/mungers/approval-handler.go]: nit: space between // and sort

File: mungegithub/mungers/submit-queue.go
Comment (philips) [inline mungegithub/mungers/submit-queue.go]: sort of a pointless comment.

**Review by philips — CHANGES_REQUESTED** [review]
Overall LGTM but some general code nits.

**Comment by philips** [comment]
I think the mechanical OWNERS changes should be in a separate PR.

**Comment by grodrigues3** [comment]
@thockin Sorry about that Tim.  This is the PR I referenced that

- gates on approved and lgtm label
- requires approval for all files in a PR 
- notifies the author (involved parties) which files have been approved and which need to be approved
- notifies a subset of APPROVERs that their approval is needed
- adds flags that control whether approval is needed

Notifying you and Phillips because you were having a very related discussion on the OWNERs File PR in kubernetes.

**Comment by philips** [comment]
cc @calebamiles can you help to review this one?

**Comment by grodrigues3** [comment]
@philips the only reason the OWNERS files are in this one too is that the code that reads "approvers" from the yaml is dependent on the new OWNERs format.  The merge should probably be atomic right?  If not, I can work on splitting it into two PRs

File: mungegithub/mungers/approval-handler.go
Comment (calebamiles) [inline mungegithub/mungers/approval-handler.go]: any reason not to do this check on `notifications` itself and possibly exit early? https://github.com/kubernetes/contrib/pull/2076/files#diff-019924d1fdee0dc57c792e7953adced4R127

**Comment by calebamiles** [comment]
looks essentially good to me although I do find it a little harder than I would like to try and grok the state machine all the mungers go through

File: mungegithub/mungers/submit-queue_test.go
Comment (grodrigues3) [inline mungegithub/mungers/submit-queue_test.go]: @calebamiles does this work?  It's used below and should NOT lead to status merged.  Should be pending with the message that there's no lgtm or approved.

**Comment by grodrigues3** [comment]
@philips @calebamiles I addressed your comments/nits.  Let me know if there's anything else.

**Comment by philips** [comment]
Overall it seemed fine and my things look addressed. LGTM

**Comment by grodrigues3** [comment]
@eparis @calebamiles are we good to merge?

File: mungegithub/features/repo-updates.go
Comment (eparis) [inline mungegithub/features/repo-updates.go]: What is this being used for now? Do we need to update/rationalize the "top of .md file assignment" stuff to this new nomenclature?

File: mungegithub/features/repo-updates.go
Comment (eparis) [inline mungegithub/features/repo-updates.go]: Not your fault, but can you fix this comment? Approvers() and Reviewers() return ALL of them, not just the closest.

File: mungegithub/mungers/approval-handler.go
Comment (eparis) [inline mungegithub/mungers/approval-handler.go]: Why does the label need to be set in order for them to be removed from the approverSet?

Can I 'disapprove' something someone else approved?

File: mungegithub/mungers/blunderbuss.go
Comment (eparis) [inline mungegithub/mungers/blunderbuss.go]: This decision seems like it should be standardized on all repos.  Should it just be "reviewers" ?

File: mungegithub/mungers/submit-queue.go
Comment (eparis) [inline mungegithub/mungers/submit-queue.go]: This would be much easier for me to read if you just made it 2 checks. And you have 2 consts above.
`noLGTM` and `noApproved`

We recheck (in the SQ) if the code was updated since the LGTM but before the other munger removed the label. Should we do the same here?

**Comment by eparis** [comment]
also update `submit-queue/www/index.html`   around line 290 we discuss the /lgtm command. We should discuss /approve

File: mungegithub/features/repo-updates.go
Comment (grodrigues3) [inline mungegithub/features/repo-updates.go]: eventually it won't be used.  it's for backwards compatibility and the markdowns

File: mungegithub/mungers/approval-handler.go
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: The comment is wrong (fixing now).  We just maintain a set of people that have approved.  If all of the people that have approved cover all the files, then the PR is marked approved.  If you cancel your approval but the set of reviewers that have approve still cover all files, the PR will still be marked approved.  In other words, people can only revoke their own approval.

File: mungegithub/mungers/blunderbuss.go
Comment (apelisse) [inline mungegithub/mungers/blunderbuss.go]: Do you mean "standardized on all mungers" ?

The question is "Do we have two flags, or do we have one in `mungegithub.go`?". Both solutions are kind of gross. Also, I don't think it really matters, because ideally this flag will be removed quickly, and is just there to ease transitioning.

File: mungegithub/mungers/submit-queue.go
Comment (apelisse) [inline mungegithub/mungers/submit-queue.go]: Much better indeed

File: mungegithub/mungers/blunderbuss.go
Comment (grodrigues3) [inline mungegithub/mungers/blunderbuss.go]: i'm also unclear on what standardized on all mungers means?

File: mungegithub/mungers/approval-handler.go
Comment (ethernetdan) [inline mungegithub/mungers/approval-handler.go]: Should we be logging deletions for debugging?

File: mungegithub/mungers/blunderbuss.go
Comment (ethernetdan) [inline mungegithub/mungers/blunderbuss.go]: No opinion this particular flag but if we find we are putting repo specific information in flags, we may consider moving it into a config file.

File: mungegithub/mungers/approval-handler.go
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Logging is already done inside `DeleteComment`

**Comment by grodrigues3** [comment]
@eparis have all your comments been addressed?  @philips already LGTMed

File: mungegithub/mungers/blunderbuss.go
Comment (grodrigues3) [inline mungegithub/mungers/blunderbuss.go]: @apelisse any comments here?  Changes because of rebase on top of upstream/master

File: mungegithub/mungers/approval-handler.go
Comment (eparis) [inline mungegithub/mungers/approval-handler.go]: This is different (not necessarily wrong) that how other mungers handle comment deletion. Generally they register a function which indicates if a comment is out of date and then the generic code takes care of deletion. See:

https://github.com/kubernetes/contrib/blob/master/mungegithub/mungers/stale-green-ci.go#L129

For a simple example.

You don't have to change it. We're already parsing the mess out of comments, but it is an inconsistency across mungers.

File: mungegithub/mungers/approval-handler.go
Comment (eparis) [inline mungegithub/mungers/approval-handler.go]: You never checked `CreatedAt != nil` before you use it here. While it shouldn't happen, the github API is not always perfect...

**Review by eparis — COMMENTED** [review]
I created https://github.com/eparis/contrib/commit/87ba9269fd6248f19073184ccfa982d08199e8ea.patch while I was trying to review this PR.

You can take it or leave it. I does fix one of my review comments. It also makes the code a lot more readable.

I take a bit of issue with some of the style here. Lines that are 1 indention deep but which run 120+ characters are hard to review. Writing like a middle schooler in nice simple statement makes review and maintenance a lot easier. Some of the time I found long lines I wasn't immediately certain what was going on I broke it down into small managable steps (the fit in my small brain).

You can either take these changes or not.

Either way you may then squash to 2 commits and then merge.

**Comment by philips** [comment]
@grodrigues3 Where are you on taking in @eparis's [final review summary](https://github.com/kubernetes/contrib/pull/2076#pullrequestreview-12341066)? Can I help?

**Comment by bgrant0607** [comment]
I would like this merged ASAP, so I would appreciate it if whatever obstacles to that could be overcome.

**Pull request merged**
